tests/phpunit/CRM/Contact/Import/Parser/ContactTest.php

```diff
@@ -2094,10 +2111,6 @@ public function getImportedContacts(): array {
   /**
    * Test that import parser will not throw error if Related Contact is not found via passed in External ID.
    *
-   * Currently fails because validation assumes the Related contact will be found.
-   * When it is later not found creating the contact via the API throws an
-   * error for missing required fields.
-   *
    * @throws \API_Exception
```

Comment:
I wrote this test because the user really doesn't know if every related contact will be found via the external_identifier in an import.  If you comment out the organization_name in the mapper the import will crash in all DUPLICATE_ ... modes.  

I don't know what the best solution is.  Requiring enough fields to create the Related Contact if not found seems a bit extreme.   Maybe treating it as an empty field would be better.  Like the code doesn't break if the phone field is blank for a imported contact.  So, if there is no Related contact to create the relationship with then don't create the relationship.  

Now that I've been so down in the details in all this.  It really is looking like the Import functionality maps and validates input fields to create an API chain.   I know you have an import api extension.  I haven't looked into it too much.  Maybe that is the direction you are already headed.  

```[{
	"resource": "/Users/darrick/buildkit/build/my-drupal-civimaster/web/sites/all/modules/civicrm/tests/phpunit/CRM/Contact/Import/Parser/ContactTest.php",
	"owner": "_generated_diagnostic_collection_name_#2",
	"severity": 8,
	"message": "CiviCRM_API3_Exception: Mandatory key(s) missing from params array: organization_name",
	"source": "PHPUnit",
	"startLineNumber": 2118,
	"startColumn": 3,
	"endLineNumber": 2118,
	"endColumn": 74,
	"relatedInformation": [
		{
			"startLineNumber": 135,
			"startColumn": 5,
			"endLineNumber": 135,
			"endColumn": 133,
			"message": "CiviCRM_API3_Exception: Mandatory key(s) missing from params array: organization_name",
			"resource": "/Users/darrick/buildkit/build/my-drupal-civimaster/web/sites/all/modules/civicrm/api/api.php"
		},
		{
			"startLineNumber": 1058,
			"startColumn": 5,
			"endLineNumber": 1058,
			"endColumn": 57,
			"message": "CiviCRM_API3_Exception: Mandatory key(s) missing from params array: organization_name",
			"resource": "/Users/darrick/buildkit/build/my-drupal-civimaster/web/sites/all/modules/civicrm/CRM/Contact/Import/Parser/Contact.php"
		},
		{
			"startLineNumber": 416,
			"startColumn": 11,
			"endLineNumber": 416,
			"endColumn": 109,
			"message": "CiviCRM_API3_Exception: Mandatory key(s) missing from params array: organization_name",
			"resource": "/Users/darrick/buildkit/build/my-drupal-civimaster/web/sites/all/modules/civicrm/CRM/Contact/Import/Parser/Contact.php"
		},
		{
			"startLineNumber": 2142,
			"startColumn": 5,
			"endLineNumber": 2142,
			"endColumn": 67,
			"message": "CiviCRM_API3_Exception: Mandatory key(s) missing from params array: organization_name",
			"resource": "/Users/darrick/buildkit/build/my-drupal-civimaster/web/sites/all/modules/civicrm/tests/phpunit/CRM/Contact/Import/Parser/ContactTest.php"
		},
		{
			"startLineNumber": 262,
			"startColumn": 7,
			"endLineNumber": 262,
			"endColumn": 32,
			"message": "CiviCRM_API3_Exception: Mandatory key(s) missing from params array: organization_name",
			"resource": "/Users/darrick/buildkit/build/my-drupal-civimaster/web/sites/all/modules/civicrm/tests/phpunit/CiviTest/CiviUnitTestCase.php"
		},
		{
			"startLineNumber": 671,
			"startColumn": 5,
			"endLineNumber": 671,
			"endColumn": 36,
			"message": "CiviCRM_API3_Exception: Mandatory key(s) missing from params array: organization_name",
			"resource": "/Users/darrick/buildkit/extern/phpunit8/phpunit8.phar"
		}
	]
}]```

Reply:
thanks for the input - I'm tying to think it all through @darrick

You are right that the goal is to ultimately map to api calls (as an aside I need to fix the api importer now that core code has changed that it relies on) - the process as I understand it is :

- the saved mapping is combined with each row in the csv to get a mapping that somewhat follows the api structure
- the validate function ensures that anything that is of the wrong data type is reported before they get to the preview screen
- the import function then does the heavy checking - but fundamentally it will skip if it doesn't have correct data to create the contact and their related contacts. Often users will run their data through for a first pass, download the fails, go through and tweak them and then upload again (I actually have a vision of them being able to do that directly in the import table at some point). So I don't think we want to require `organization_name ` in case there is no `external_identifer` match found - since their process might be to see how many go in the first time & then refine

I'm trying to put togehter a full picture of expecations to test against now - here is my first cut. I'll transfer it somewhere once I've brain dumped a bit


| Mode      | Available fields |Behaviour Main contact|Behaviour related contact
| ----------- | ----------- | ----------- |----------- |
| Skip     | external identifier, dedupe rule fields  , not id |If contact found by ext id or dedupe rule then skip, |if related contact found by dedupe rule, then use it,or create related otherwise
| No match   | id, external identifier, dedupe rules      | if external identifier provided and it matches an existing contact or an existing contact can be found using rules, skip. No idea the point ot the id|if a single match can be found using id, ext id, rules then updated it,otherwise create, if possible|
| Update| id, external identifier, dedupe rules | if a single match can be found using id, ext id, rules then updated it, otherwise skip|if a single match can be found using id, ext id, rules then updated it,otherwise create, if possible|
|Fill|id, external identifier, dedupe rules | if a single match can be found using id, ext id, rules then fill in missing data, otherwise skip|if a single match can be found using id, ext id, rules then updated it,otherwise create, if possible|

Reply:
So there are two things you can configure for the main contact that you can't configure for the related....

- action if not found (currently create)
- rule to use on related contact (currently defaults to Unsupervised for the type - I *think* that is even if the type is the same as the main contact but would have to guess)

I am wondering about adding an option to the first screen

'Action to take if related contact is not found' 
- create the contact, if possible, otherwise skip the whole row
- create the contact, if possible, otherwise skip the related contact
- skip the related contact

Reply:
What happens if the contact isn't found when importing a contribution or activity (Just did a quick test, it returns "CRM_Import_Parser::ERROR")?   I think contact import should do the same to keep things consistent.  Currently it just crashes.  So, that is a bug.

Then I think being able to select a dedupe rule for the target contact when importing contacts, contributions, activities, etc.  would be super beneficial.   As the dedupe rule may depend on the source of the data.

Once that is all sorted a "triage" wizard can be worked on.  Currently that is like you said.  Download the errors, sort it out manually and try again.  Could also provide a hook so a programmer can decide to just create the contact and rerun the import on the row.

Reply:
Currently it just crashes. So, that is a bug. - so you are saying there is a fatal error when not found now  rather than updating the import table with an error & moving onto the next one? Will check that out

Reply:
Yes.  I'm seeing if I can fix that right now with this branch: https://github.com/darrick/civicrm-core/tree/deprecated_contact_check_params

Reply:
Can you give me steps to replicate?


I just tried

![image](https://user-images.githubusercontent.com/336308/171524195-c1610735-0b94-466b-96b3-f8e646722c1c.png)

and got

![image](https://user-images.githubusercontent.com/336308/171524226-41bb440d-a206-4b33-9c8b-d74bdfd5a107.png)


![image](https://user-images.githubusercontent.com/336308/171524261-4b4595b6-3ba8-454f-b5df-67d79a98b4e4.png)


I think from the e-notice the exception was maybe caught in this block - which I think we want to NOT happen - so I'm gonna check that out - but is that the error you are describing?
```
      catch (CiviCRM_API3_Exception $e) {
        // When we catch errors here we are not adding to the errors array - mostly
        // because that will become obsolete once https://github.com/civicrm/civicrm-core/pull/23292
        // is merged and this will replace it as the main way to handle errors (ie. update the table
        // and move on).
        $this->setImportStatus((int) $values[count($values) - 1], 'ERROR', $e->getMessage());
      }
```

Reply:
@darrick  - I just put up https://github.com/civicrm/civicrm-core/pull/23660